collector.c

```diff
@@ -120,7 +120,16 @@ probe_waits(const bool write_history, const bool write_profile)
 	if (write_history)
 		LWLockAcquire(pgws_history_lock, LW_EXCLUSIVE);
 
-	/* Iterate PGPROCs under shared lock */
+	/*
+	 * Iterate PGPROCs under shared lock.
+	 *
+	 * TODO:
+	 * ProcArrayLock is heavy enough and in current case we might perform the
+	 * non-trivial deallocation routine for profile hash table under this lock.
+	 * Therefore to reduce possible contention it's worth to segregate the logic
+	 * of PGPROCs iteration under ProcArrayLock and storing results to profile
+	 * and/or history under corresponding another lock.
+	 */
 	LWLockAcquire(ProcArrayLock, LW_SHARED);
```

Comment:
Why do we take ProcArrayLock anyway? What concurrent operations does it prevent?
I don't think that blocking ProcArrayAdd and ProcArrayRemove buys us anything. We are not iterating over the procArray, but over allProcs. And we don't need the lock to read PGPROC.pid and wait_event_info.

Reply:
Yes, you are right. In current case when we verify slot validity also via non-zero `wait_event_info` value we might omit acquisition of this lock. This is mostly inherited from previous implementation.